### DISCUSSION TITLE
CoreDNS : How to call forward plugin from another plugin

### DISCUSSION THREAD

**Discussion by Sharaf5** [coredns/coredns#6811] [Q&A]
I'm trying to do some sort of plugin "for captive portals" that forward on some conditions and get records from other data sources in other conditions where I cant relay on the CoreFile plugin sequence to call next one

is there a way to call it with it's options from other plugins

I'm handling forward like this now, but it's doesnt seem to be the proper way
```
func (c *Captive) handleForward(ctx context.Context, w dns.ResponseWriter, r *dns.Msg, forward string, RDN string) (int, error) {
	if forward == "" {
		return dns.RcodeServerFailure, fmt.Errorf("Err30: No forward destination configured")
	}

	client := &dns.Client{}
	forwardAddr := net.JoinHostPort(forward, "53")

	if RDN != "" {
		// Captive forward
		rClone := c.CloneMsg(r)
		for i := range rClone.Question { rClone.Question[i].Name = RDN }

		capAddr := net.JoinHostPort(forward, "53")
		capResp, _, err := client.Exchange(rClone, capAddr)
		if err != nil {
			c.logMsg("handleForward", "Err15", "Failed to forward query to captive server: ", err)
			return dns.RcodeServerFailure, err
		}
		c.logMsg("handleForward", "Dbg13", "Captive Forward Response received: rClone :: ", rClone.Question, rClone.Answer, ", captiveResp :: ", capResp.Question, capResp.Answer)

		if len(capResp.Answer) < 1 {
			return dns.RcodeServerFailure, fmt.Errorf("Forward response empty")
		}
		for i := range capResp.Answer {
			capResp.Answer[i].Header().Name =  r.Question[0].Name
			capResp.Answer[i].Header().Ttl = 0
		}

		return c.generateDNSMessage(w, r, capResp.Answer, r.Question)
	} else {
		// Normal forward
		resp, _, err := client.Exchange(r, forwardAddr)
		if err != nil {
			c.logMsg("handleForward", "Err15", "Failed to forward query: ", err)
			return dns.RcodeServerFailure, err
		}
		c.logMsg("handleForward", "Dbg13", "Normal Forward Response received: ", resp.Question, resp.Answer)
		return c.generateDNSMessage(w, r, resp.Answer, resp.Question)
	}
	
	return dns.RcodeServerFailure, fmt.Errorf("Failed to Forward request")
}
```

now I want to replace the simple `client.Exchange(r, forwardAddr)` I have with more fully featured forwarder and I think yours should fit all needs

**Comment by chrisohaver** [comment]
The _view_ plugin does conditional routing.

**Reply by Sharaf5** [reply]
it's a captive portal full solution in one plugin

plugin include dynamic view handling from db, serving view data either from db, zone file, forwarding or serving default captive data, identifing authenticated uses each with his view. and some extra dns types like from those that "github.com/miekg/dns" could know it's type but not serving data
```
// All types : None A NS MD MF CNAME SOA MB MG MR NULL PTR HINFO MINFO MX TXT RP AFSDB X25 ISDN RT NSAPPTR SIG KEY PX GPOS AAAA LOC NXT EID NIMLOC SRV ATMA NAPTR KX CERT DNAME OPT APL DS SSHFP IPSECKEY RRSIG NSEC DNSKEY DHCID NSEC3 NSEC3PARAM TLSA SMIMEA HIP NINFO RKEY TALINK CDS CDNSKEY OPENPGPKEY CSYNC ZONEMD SVCB HTTPS SPF UINFO UID GID UNSPEC NID L32 L64 LP EUI48 EUI64 NXNAME URI CAA AVC AMTRELAY TKEY TSIG alid IXFR AXFR MAILB MAILA ANY TA DLV Reserved
```

plugin is meant to be the only one loaded and serving data, any other behavior or plugin should be loaded on configeration set in db
so it really act as a captive portal dns server
- identifing view dynamically
- serving from db
- serving from file if configuerd
- serving forwarded requests if configuerd
- white/black list sites dynamically
- serving portal sites (for non-auth client, no data response, black listed) and some `local.` sites like coffies and airports
and some other minor features

I know most of these features has thier plugins but all should be declaired statically so every configuration should reload manually or by complex script
this plugin should define rolles when other plugins called ( if there is a way to call plugins from other plugins ) so the only static thing is that zone files untill

the part above is two forward scnarios :
- normal forward : only forward request
- captive forward : forward any request to the forward address as if was the captive portal question
it's not conditional routing but just forwarding but for the true question or the captive site if it was puplic one
Of course it's not the best way as dns queries cant give http redirect response so it's the only solution for now
